File: Leetcode 0144 Binary Tree Preorder Traversal [medium] [tree]/Binary Tree Preorder Traversal [iterative].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        
        List<Integer> preorder = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        
        TreeNode cur = root;
        
        if (root == null) {
            return preorder;
        }
        
        while (!stack.empty() || cur != null) {
            
            while (cur != null) {
                preorder.add(cur.val);
                stack.push(cur);
                cur = cur.left;
            }
            
            if (!stack.empty()) {
                cur = stack.pop();
                cur = cur.right;
            }
        }
        
        return preorder;
        
    }
}

File: Leetcode 0144 Binary Tree Preorder Traversal [medium] [tree]/Binary Tree Preorder Traversal [iterative].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    let preorder = [];
    let stack = [];
    
    if (root === null) {
        return preorder;
    }
    
    let cur = root;
    while (stack.length !== 0 || cur !== null) {
        while (cur !== null) {
            preorder.push(cur.val);
            stack.push(cur);
            cur = cur.left;
        }
        
        if (stack.length !== 0) {
            cur = stack.pop();
            cur = cur.right;
        }
    }
    
    return preorder;
};